### PR TITLE
[FLINK-8058][Queryable State]Queryable state should check types

### PR DESCRIPTION
**Pull request by klion26**
## What is the purpose of the change

Add type checks(state type and state value type) in server side

## Verifying this change

This change added tests and can be verified as follows:

  - *Added mismatch state type and state value type test*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature?  no

### CODE REVIEW

**Comment by klion26** [comment]
@kl0u @twalthr could you please help review this pr?
In other word, I have one more question: If I need to add a test that`KvStateSerializer#deserializeStateDescriptor` throws `ClassNotFoundException`, if yes, if there any test I can refer.

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/KvStateServerHandlerTest.java [hunk 1]
```diff
@@ -676,12 +736,15 @@ public void testChunkedResponse() throws Exception {
...
 				VoidNamespaceSerializer.INSTANCE);
+		byte[] serializedStateDescriptor = KvStateSerializer.serializedStateDescriptor(desc);
 
 		long requestId = Integer.MAX_VALUE + 182828L;
 
 		assertTrue(registryListener.registrationName.equals("vanilla"));
 
-		KvStateInternalRequest request = new KvStateInternalRequest(registryListener.kvStateId, serializedKeyAndNamespace);
+		KvStateInternalRequest request = new KvStateInternalRequest(registryListener.kvStateId,
+																	serializedKeyAndNamespace,
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/KvStateServerHandlerTest.java]: format the code

File: flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/KvStateServerHandlerTest.java [hunk 2]
```diff
@@ -259,12 +261,15 @@ public void testQueryUnknownKey() throws Exception {
...
 				VoidNamespaceSerializer.INSTANCE);
+		byte[] serializedStateDescriptor = KvStateSerializer.serializedStateDescriptor(desc);
 
 		long requestId = Integer.MAX_VALUE + 22982L;
 
 		assertTrue(registryListener.registrationName.equals("vanilla"));
 
-		KvStateInternalRequest request = new KvStateInternalRequest(registryListener.kvStateId, serializedKeyAndNamespace);
+		KvStateInternalRequest request = new KvStateInternalRequest(registryListener.kvStateId,
+																	serializedKeyAndNamespace,
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-runtime/src/test/java/org/apache/flink/queryablestate/network/KvStateServerHandlerTest.java]: format the code

File: flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java [hunk 3]
```diff
@@ -81,6 +83,14 @@ public KvStateServerHandler(
 			} else {
 				byte[] serializedKeyAndNamespace = request.getSerializedKeyAndNamespace();
 
+				StateDescriptor<?, ?> requestStateDescriptor = KvStateSerializer.deserializeStateDescriptor(request.getSerializedStateDescriptor());
+				StateDescriptor<?, ?> registStateDescriptor = kvState.getStateDescriptor();
+
+				Preconditions.checkArgument(requestStateDescriptor.getType().equals(registStateDescriptor.getType()),
+					"State type mismatch, need[" + registStateDescriptor.getType() + "] gotten[" + requestStateDescriptor.getType() + "]");
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: use string format looks better~

File: flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java [hunk 4]
```diff
@@ -81,6 +83,14 @@ public KvStateServerHandler(
...
+				Preconditions.checkArgument(requestStateDescriptor.getSerializer().equals(registStateDescriptor.getSerializer()),
+					"State value serializer mismatch, need [" + registStateDescriptor.getSerializer() + "] gotten[" + requestStateDescriptor.getSerializer() + "]");
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-runtime/src/main/java/org/apache/flink/queryablestate/server/KvStateServerHandler.java]: use string format looks better~

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/messages/KvStateRequest.java [hunk 5]
```diff
@@ -70,17 +73,23 @@ public int getKeyHashCode() {
...
 
 		// JobID + stateName + sizeOf(stateName) + hashCode + keyAndNamespace + sizeOf(keyAndNamespace)
+		// + stateDescriptor + sizeOf(stateDescriptor)
 		final int size =
 				JobID.SIZE +
 				serializedStateName.length + Integer.BYTES +
 				Integer.BYTES +
-				serializedKeyAndNamespace.length + Integer.BYTES;
+				serializedKeyAndNamespace.length + Integer.BYTES
+				+ serializedStateDescriptor.length + Integer.BYTES;
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/messages/KvStateRequest.java]: put the "+" behind the previous line looks better to me~

File: flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/client/state/serialization/KvStateSerializer.java [hunk 6]
```diff
@@ -264,4 +269,39 @@
...
+	 * @param stateDescriptor the value will be serialized.
+	 *
+	 * @return The serialized values
+	 * @throws IOException On failure during serialization
+	 */
+	public static byte[] serializedStateDescriptor(StateDescriptor<?, ?> stateDescriptor) throws IOException {
+		ByteArrayOutputStream bos = new ByteArrayOutputStream();
+		ObjectOutputStream out = new ObjectOutputStream(bos);
+		out.writeObject(stateDescriptor);
+		out.close();
```
Comment (yanghua) [inline flink-queryable-state/flink-queryable-state-client-java/src/main/java/org/apache/flink/queryablestate/client/state/serialization/KvStateSerializer.java]: shall we use `try-with-resource` to make sure the stream can be closed?

**Comment by klion26** [comment]
@yanghua thank you for helping review, I have updated the code~

**Comment by azagrebin** [comment]
hi @klion26, do I understand correctly that the idea now is that the state descriptor is serialised and sent for every request? If so it is probably too expensive to do it every time. The check could happen only once.

**Comment by azagrebin** [comment]
@klion26, I suggest we get back to Jira issue and discuss the approach. In general, if design is firstly described in Jira issue and there is a feedback from a committer, it simplifies and speeds up the review and merging. I would close this PR and open another one when there is an agreement in Jira issue about the implementation approach.

**Comment by klion26** [comment]
@azagrebin ok, i'll close the PR.

**Pull request closed without merging**